networkd/network_server.ml

```diff
@@ -130,13 +130,26 @@ module Sriov = struct
       else
         Ok Modprobe_successful_requires_reboot
     | None ->
-      debug "%s SR-IOV on a device: %s via sysfs" op dev;
+      (* enable: try sysfs interface to set numfvs = maxvfs. if fails, but vfs are enabled, assume manual configuration.
+         disable: Net.Sriov.disable will not be called for manually configured interfaces, as determined by `require_operation_on_pci_device` *)
+      (* let Ok(maxvfs) = Sysfs.get_sriov_maxvfs dev in *)
+      if enable then
       begin
-        if enable then Sysfs.get_sriov_maxvfs dev
-        else Sysfs.unbind_child_vfs dev >>= fun () -> Ok 0
-      end >>= fun numvfs ->
-      Sysfs.set_sriov_numvfs dev numvfs >>= fun _ ->
-      Ok Sysfs_successful
+        let present_numvfs = Sysfs.get_sriov_numvfs dev in
+        match (Sysfs.get_sriov_maxvfs dev >>= fun maxvfs -> maxvfs |> Sysfs.set_sriov_numvfs dev), present_numvfs
+        with
+        | Ok _ , _ -> debug "%s SR-IOV on a device: %s via sysfs" op dev; Ok Sysfs_successful
+        | _, _ -> debug "SR-IOV/VFs %sd manually on device: %s" op dev; Ok Manual_successful
```

Comment:
You are not actually matching on `present_numvfs`. Don't you need at least `| _ present -> when present > 0`? Or better, check that it is equal to `maxvfs`?

Reply:
You are absolutely right! When adapting the code I was more amazed by the fact that I may integrate the exception handling in the match, and forgot to think twice about the actual semantics.
The code now does something very different and it just tested fine on my machine, as currently enabled VFs and maximum number of VFs is equal.
I am currently thinking about a way to rewrite the code with my new knowledge about matching of exceptions, but for now did not come to a conclusion. I think the old way of implementing this + the "style-fixes" will be the way to go.
I hope to come to a conclusion within the next hour.

Reply:
I don't think it is a problem is the number of VFs is smaller than the max. At least, I hope that we do the accounting, when allocating VFs to VMs, based on the current number rather than the max. As long as the number is larger than zero. So the `when` clause I suggested above may be enough?

Reply:
And to actually reply to your question:
> Don't you need at least `| _, present -> when present > 0`? Or better, check that it is equal to `maxvfs`?

I would need to check if it is `>0` or `<>0`. Problem here is, that the code to set the new number of VFs is expected to throw. I have to match for `exception` and `> 0`, or like in the last version, match `exception, 0` first and know that anything matched afterwards can only be `exception` and `> 0`. This was the reason I built the `Ok()` and `Error()` returning try-catch-construct

Reply:
> `| _, present -> when present > 0`
> So the `when` clause I suggested above may be enough?

The attempt to set a new number of VFs may return with `Error _`, when a failure is handled internally by the called methods, or may just throw. Both, `exception` and `Error`, are acceptable in the case of `present_numvfs > 0`.
As far as I understand, the wildcard `_` does not match exceptions. But trying to match multi case on `| exception _ | Error _ when present_numvfs > 0 -> ...` is unfortunately not supported either (no mixing of value and exception patterns when using when-guard).
Including the exception in the match is possible, but for now I cannot find a solution that is actually cleaner and more concise than the original variant. Combining `Error` and `exception` into one match case by catching exceptions beforehand, effectively avoids code duplication.

It is possible to get rid of the try-catch and matching a tuple, but the match is not any more readable, due to the reason given above. Example:
```
| Ok _ -> debug [...]; Ok
| Error _ when present_numvfs > 0 -> debug [...]; Ok Manual_successful
| exception _ when present_numvfs > 0 -> debug [...]; Ok Manual_successful
| Error err -> Error err
| exception e -> let msg = [...]; Error(Other, msg)
```

Reply:
It is correct that a single pattern either matches a normal value or an exception, but never both. I still think a collection of patterns is easier to reason about and easier to extend than if-then-else or try-with chains. So I do like the above but would not insist on it.

Reply:
In that case, my take at the implementation with the least redundancy would be as follows:
```ocaml
let man_successful =
    debug "SR-IOV/VFs %sd manually on device: %s" op dev;
    Manual_successful
in
if enable then
begin
    let present_numvfs = Sysfs.get_sriov_numvfs dev in
    match Sysfs.get_sriov_maxvfs dev >>= fun maxvfs -> maxvfs |> Sysfs.set_sriov_numvfs dev
    with
    | Ok _ -> debug "%s SR-IOV on a device: %s via sysfs" op dev; Ok Sysfs_successful
    | Error _     when present_numvfs > 0 -> Ok man_successful
    | exception _ when present_numvfs > 0 -> Ok man_successful
    | Error err -> Error err
    | exception e ->
        let msg = Printf.sprintf "Error: trying sysfs SR-IOV interface failed with exception %s on device: %s" (Printexc.to_string e) dev in
        Error(Other, msg)
end
```

should be equivalent to following code with doublings, but without introduction of an additional function:
```ocaml
if enable then
begin
    let present_numvfs = Sysfs.get_sriov_numvfs dev in
    match Sysfs.get_sriov_maxvfs dev >>= fun maxvfs -> maxvfs |> Sysfs.set_sriov_numvfs dev
    with
    | Ok _ -> debug "%s SR-IOV on a device: %s via sysfs" op dev; Ok Sysfs_successful
    | Error _     when present_numvfs > 0 -> debug "SR-IOV/VFs %sd manually on device: %s" op dev; Ok Manual_successful
    | exception _ when present_numvfs > 0 -> debug "SR-IOV/VFs %sd manually on device: %s" op dev; Ok Manual_successful
    | Error err -> Error err
    | exception e ->
        let msg = Printf.sprintf "Error: trying sysfs SR-IOV interface failed with exception %s on device: %s" (Printexc.to_string e) dev in
        Error(Other, msg)
end
```

or, using try catch:
```ocaml
begin
    let present_numvfs = Sysfs.get_sriov_numvfs dev in
    match
    ( try Sysfs.get_sriov_maxvfs dev >>= fun (maxvfs) -> maxvfs |> Sysfs.set_sriov_numvfs dev
        with e ->
        let msg = Printf.sprintf "Error: trying sysfs SR-IOV interface failed with exception %s on device: %s" (Printexc.to_string e) dev in
        Error(Other, msg) ),
    present_numvfs
    with
    | Ok _ , _ -> debug "%s SR-IOV on a device: %s via sysfs" op dev; Ok Sysfs_successful
    | Error(err), 0 -> Error(err)
    | _, _ -> debug "SR-IOV/VFs %sd manually on device: %s" op dev; Ok Manual_successful
end
```

if there are any preferences, tell me and I can commit the code

Reply:
@lindig @robhoes : any preferences regarding style of the implementation?

Reply:
I like the first option of the three, if that helps you unlock the work :)